File: freight-triage-polyglot-ledger-recovery/environment/native/src/selftest.cpp
#include "freight/selftest.h"

#include <cstdio>
#include <map>
#include <string>

#include "freight/codecs.h"
#include "freight/formats.h"
#include "freight/hashes.h"
#include "freight/normalize.h"
#include "freight/sha256.h"
#include "freight/stats.h"
#include "freight/tables.h"

namespace freight {
namespace {

const int kProbeCount = 64;
const int kSeriesCount = 24;
const int kSeriesLength = 17;
const int kRecordCount = 40;

std::string hex64(unsigned long long value) {
  char buffer[32];
  std::snprintf(buffer, sizeof(buffer), "%016llx", value);
  return std::string(buffer);
}

unsigned long long foldFnv1a64(unsigned long long state, const std::string& text) {
  for (size_t i = 0; i < text.size(); ++i) {
    state ^= static_cast<unsigned char>(text[i]);
    state *= 1099511628211ULL;
  }
  return state;
}

}  // namespace

unsigned int probeMix32(unsigned int seed) {
  unsigned int x = seed;
  x ^= x >> 16;
  x *= 0x7FEB352Du;
  x ^= x >> 15;
  x *= 0x846CA68Bu;
  x ^= x >> 16;
  return x;
}

std::string probeString(int index) {
  char head[32];
  std::snprintf(head, sizeof(head), "FRT-%04d-", index);
  std::string out(head);
  const int tail = index % 11;
  for (int k = 0; k < tail; ++k) {
    out.push_back(static_cast<char>('a' + ((index * 7 + k * 3) % 26)));
  }
  return out;
}

std::vector<long long> probeSeries(int series) {
  std::vector<long long> values;
  values.reserve(kSeriesLength);
  for (int k = 0; k < kSeriesLength; ++k) {
    values.push_back(3 + static_cast<long long>(probeMix32(static_cast<unsigned int>(series * 977 + k * 31)) % 4093u));
  }
  return values;
}

std::vector<ProbeRecord> probeRecords() {
  std::vector<ProbeRecord> records;
  records.reserve(kRecordCount);
  for (int index = 0; index < kRecordCount; ++index) {
    unsigned int m = probeMix32(static_cast<unsigned int>(index * 131 + 17));
    char id[16];
    std::snprintf(id, sizeof(id), "RC-%03d", index);
    ProbeRecord record;
    record.recordId = id;
    record.laneIndex = m % 360u;
    record.massKg = 50 + static_cast<long long>((m >> 7) % 48000u);
    record.priority = static_cast<long long>((m >> 3) % 5u);
    record.hazmatClass = static_cast<long long>((m >> 11) % 9u);
    record.sealLength = 6 + static_cast<long long>((m >> 17) % 7u);
    records.push_back(record);
  }
  return records;
}

Json buildSelftestReport() {
  std::map<std::string, std::map<std::string, std::string> > families;

  const std::vector<HashAlgorithm>& hashes = hashRegistry();
  for (size_t i = 0; i < hashes.size(); ++i) {
    unsigned long long folded = 14695981039346656037ULL;
    for (int p = 0; p < kProbeCount; ++p) {
      folded = foldFnv1a64(folded, hex64(hashes[i].apply(probeString(p))));
    }
    families["hash"][hashes[i].name] = hex64(folded);
  }

  const std::vector<CodecAlgorithm>& codecs = codecRegistry();
  for (size_t i = 0; i < codecs.size(); ++i) {
    unsigned long long folded = 14695981039346656037ULL;
    bool roundTrip = true;
    for (int p = 0; p < kProbeCount; ++p) {
      const std::string probe = probeString(p);
      const std::string encoded = codecs[i].encode(probe);
      folded = foldFnv1a64(folded, encoded);
      if (codecs[i].decode(encoded) != probe) {
        roundTrip = false;
      }
    }
    if (!roundTrip) {
      folded ^= 0xDEADBEEFCAFEF00DULL;
    }
    families["codec"][codecs[i].name] = hex64(folded);
  }

  const std::vector<StatKernel>& kernels = statRegistry();
  for (size_t i = 0; i < kernels.size(); ++i) {
    unsigned long long folded = 14695981039346656037ULL;
    for (int s = 0; s < kSeriesCount; ++s) {
      char buffer[32];
      std::snprintf(buffer, sizeof(buffer), "%lld", kernels[i].apply(probeSeries(s)));
      folded = foldFnv1a64(folded, buffer);
    }
    families["stats"][kernels[i].name] = hex64(folded);
  }

  const std::vector<TriageRule>& rules = ruleRegistry();
  const std::vector<ProbeRecord> records = probeRecords();
  for (size_t i = 0; i < rules.size(); ++i) {
    unsigned long long folded = 14695981039346656037ULL;
    for (size_t r = 0; r < records.size(); ++r) {
      folded = foldFnv1a64(folded, rules[i].apply(records[r]) ? "1" : "0");
    }
    families["rules"][rules[i].name] = hex64(folded);
  }

  const std::vector<Formatter>& formatters = formatterRegistry();
  for (size_t i = 0; i < formatters.size(); ++i) {
    unsigned long long folded = 14695981039346656037ULL;
    for (int s = 0; s < kSeriesCount; ++s) {
      const std::vector<long long> series = probeSeries(s);
      for (size_t v = 0; v < series.size(); ++v) {
        folded = foldFnv1a64(folded, formatters[i].apply(series[v]));
        folded = foldFnv1a64(folded, formatters[i].apply(-series[v]));
      }
    }
    families["format"][formatters[i].name] = hex64(folded);
  }

  const std::vector<Normalizer>& normalizers = normalizerRegistry();
  for (size_t i = 0; i < normalizers.size(); ++i) {
    unsigned long long folded = 14695981039346656037ULL;
    for (int p = 0; p < kProbeCount; ++p) {
      folded = foldFnv1a64(folded, normalizers[i].apply(probeString(p)));
    }
    families["norm"][normalizers[i].name] = hex64(folded);
  }

  {
    unsigned long long folded = 14695981039346656037ULL;
    const std::vector<LaneTableRow>& rows = laneTableRows();
    for (size_t i = 0; i < rows.size(); ++i) {
      folded = foldFnv1a64(folded, laneTableCanonical(rows[i]));
    }
    families["tables"]["lane"] = hex64(folded);
  }
  {
    unsigned long long folded = 14695981039346656037ULL;
    const std::vector<CarrierTableRow>& rows = carrierTableRows();
    for (size_t i = 0; i < rows.size(); ++i) {
      folded = foldFnv1a64(folded, carrierTableCanonical(rows[i]));
    }
    families["tables"]["carrier"] = hex64(folded);
  }
  {
    unsigned long long folded = 14695981039346656037ULL;
    const std::vector<CommodityTableRow>& rows = commodityTableRows();
    for (size_t i = 0; i < rows.size(); ++i) {
      folded = foldFnv1a64(folded, commodityTableCanonical(rows[i]));
    }
    families["tables"]["commodity"] = hex64(folded);
  }
  {
    unsigned long long folded = 14695981039346656037ULL;
    const std::vector<TariffTableRow>& rows = tariffTableRows();
    for (size_t i = 0; i < rows.size(); ++i) {
      folded = foldFnv1a64(folded, tariffTableCanonical(rows[i]));
    }
    families["tables"]["tariff"] = hex64(folded);
  }
  {
    unsigned long long folded = 14695981039346656037ULL;
    const std::vector<ZoneTableRow>& rows = zoneTableRows();
    for (size_t i = 0; i < rows.size(); ++i) {
      folded = foldFnv1a64(folded, zoneTableCanonical(rows[i]));
    }
    families["tables"]["zone"] = hex64(folded);
  }
  {
    unsigned long long folded = 14695981039346656037ULL;
    const std::vector<HazmatTableRow>& rows = hazmatTableRows();
    for (size_t i = 0; i < rows.size(); ++i) {
      folded = foldFnv1a64(folded, hazmatTableCanonical(rows[i]));
    }
    families["tables"]["hazmat"] = hex64(folded);
  }

  Sha256 hasher;
  Json familyJson = Json::object();
  for (std::map<std::string, std::map<std::string, std::string> >::const_iterator family =
           families.begin();
       family != families.end(); ++family) {
    Json bucket = Json::object();
    for (std::map<std::string, std::string>::const_iterator entry = family->second.begin();
         entry != family->second.end(); ++entry) {
      bucket[entry->first] = Json(entry->second);
      hasher.update(family->first + "|" + entry->first + "|" + entry->second + "\n");
    }
    familyJson[family->first] = bucket;
  }

  Json report = Json::object();
  report["digest"] = Json(hasher.hexDigest());
  report["families"] = familyJson;
  report["generator"] = Json(std::string("cpp"));
  report["probe_count"] = Json(static_cast<long long>(kProbeCount));
  report["record_count"] = Json(static_cast<long long>(kRecordCount));
  report["schema_version"] = Json(std::string("freight-selftest/2"));
  report["series_count"] = Json(static_cast<long long>(kSeriesCount));
  return report;
}

}  // namespace freight

File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/selftest/selftest.go
// Package selftest builds the cross language conformance report.
package selftest

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sort"

	"freight/reconcile/internal/codecx"
	"freight/reconcile/internal/formatx"
	"freight/reconcile/internal/hashx"
	"freight/reconcile/internal/normx"
	"freight/reconcile/internal/rulex"
	"freight/reconcile/internal/statx"
	"freight/reconcile/internal/tablex"
)

const (
	probeCount   = 64
	seriesCount  = 24
	seriesLength = 17
	recordCount  = 40
)

const fnvOffset uint64 = 14695981039346656037
const fnvPrime uint64 = 1099511628211

func fold(state uint64, data []byte) uint64 {
	value := state
	for _, b := range data {
		value ^= uint64(b)
		value *= fnvPrime
	}
	return value
}

func foldString(state uint64, text string) uint64 {
	return fold(state, []byte(text))
}

func hex64(value uint64) string {
	return fmt.Sprintf("%016x", value)
}

// Mix32 is the shared deterministic probe mixer.
func Mix32(seed uint32) uint32 {
	x := seed
	x ^= x >> 16
	x *= 0x7FEB352D
	x ^= x >> 15
	x *= 0x846CA68B
	x ^= x >> 16
	return x
}

// ProbeString builds probe number index.
func ProbeString(index int) string {
	buffer := &bytes.Buffer{}
	fmt.Fprintf(buffer, "FRT-%04d-", index)
	for k := 0; k < index%11; k++ {
		buffer.WriteByte(byte('a' + ((index*7 + k*3) % 26)))
	}
	return buffer.String()
}

// ProbeSeries builds integer series number series.
func ProbeSeries(series int) []int64 {
	values := make([]int64, 0, seriesLength)
	for k := 0; k < seriesLength; k++ {
		m := Mix32(uint32(series*977 + k*31))
		values = append(values, 3+int64(m%4093))
	}
	return values
}

// ProbeRecords builds the triage record corpus.
func ProbeRecords() []rulex.Record {
	records := make([]rulex.Record, 0, recordCount)
	for index := 0; index < recordCount; index++ {
		m := Mix32(uint32(index*131 + 17))
		records = append(records, rulex.Record{
			RecordID:    fmt.Sprintf("RC-%03d", index),
			LaneIndex:   int64(m % 360),
			MassKg:      50 + int64((m>>7)%48000),
			Priority:    int64((m >> 3) % 5),
			HazmatClass: int64((m >> 11) % 9),
			SealLength:  6 + int64((m>>17)%7),
		})
	}
	return records
}

// Build assembles the Go conformance report.
func Build() map[string]any {
	families := map[string]map[string]string{}

	hashes := map[string]string{}
	for _, algorithm := range hashx.Registry() {
		folded := fnvOffset
		for p := 0; p < probeCount; p++ {
			folded = foldString(folded, hex64(algorithm.Apply([]byte(ProbeString(p)))))
		}
		hashes[algorithm.Name] = hex64(folded)
	}
	families["hash"] = hashes

	codecs := map[string]string{}
	for _, codec := range codecx.Registry() {
		folded := fnvOffset
		roundTrip := true
		for p := 0; p < probeCount; p++ {
			probe := []byte(ProbeString(p))
			encoded := codec.Encode(probe)
			folded = fold(folded, encoded)
			if !bytes.Equal(codec.Decode(encoded), probe) {
				roundTrip = false
			}
		}
		if !roundTrip {
			folded ^= 0xDEADBEEFCAFEF00D
		}
		codecs[codec.Name] = hex64(folded)
	}
	families["codec"] = codecs

	stats := map[string]string{}
	for _, kernel := range statx.Registry() {
		folded := fnvOffset
		for s := 0; s < seriesCount; s++ {
			folded = foldString(folded, fmt.Sprintf("%d", kernel.Apply(ProbeSeries(s))))
		}
		stats[kernel.Name] = hex64(folded)
	}
	families["stats"] = stats

	rules := map[string]string{}
	records := ProbeRecords()
	for _, rule := range rulex.Registry() {
		folded := fnvOffset
		for _, record := range records {
			if rule.Apply(record) {
				folded = foldString(folded, "1")
			} else {
				folded = foldString(folded, "0")
			}
		}
		rules[rule.Name] = hex64(folded)
	}
	families["rules"] = rules

	formats := map[string]string{}
	for _, formatter := range formatx.Registry() {
		folded := fnvOffset
		for s := 0; s < seriesCount; s++ {
			for _, value := range ProbeSeries(s) {
				folded = foldString(folded, formatter.Apply(value))
				folded = foldString(folded, formatter.Apply(-value))
			}
		}
		formats[formatter.Name] = hex64(folded)
	}
	families["format"] = formats

	norms := map[string]string{}
	for _, normalizer := range normx.Registry() {
		folded := fnvOffset
		for p := 0; p < probeCount; p++ {
			folded = foldString(folded, normalizer.Apply(ProbeString(p)))
		}
		norms[normalizer.Name] = hex64(folded)
	}
	families["norm"] = norms

	tables := map[string]string{}
	folded := fnvOffset
	for _, row := range tablex.LaneRows() {
		folded = foldString(folded, row.Canonical())
	}
	tables["lane"] = hex64(folded)
	folded = fnvOffset
	for _, row := range tablex.CarrierRows() {
		folded = foldString(folded, row.Canonical())
	}
	tables["carrier"] = hex64(folded)
	folded = fnvOffset
	for _, row := range tablex.CommodityRows() {
		folded = foldString(folded, row.Canonical())
	}
	tables["commodity"] = hex64(folded)
	folded = fnvOffset
	for _, row := range tablex.TariffRows() {
		folded = foldString(folded, row.Canonical())
	}
	tables["tariff"] = hex64(folded)
	folded = fnvOffset
	for _, row := range tablex.ZoneRows() {
		folded = foldString(folded, row.Canonical())
	}
	tables["zone"] = hex64(folded)
	folded = fnvOffset
	for _, row := range tablex.HazmatRows() {
		folded = foldString(folded, row.Canonical())
	}
	tables["hazmat"] = hex64(folded)
	families["tables"] = tables

	familyNames := make([]string, 0, len(families))
	for name := range families {
		familyNames = append(familyNames, name)
	}
	sort.Strings(familyNames)

	hasher := sha256.New()
	document := map[string]any{}
	for _, familyName := range familyNames {
		entries := families[familyName]
		names := make([]string, 0, len(entries))
		for name := range entries {
			names = append(names, name)
		}
		sort.Strings(names)
		bucket := map[string]any{}
		for _, name := range names {
			bucket[name] = entries[name]
			hasher.Write([]byte(familyName + "|" + name + "|" + entries[name] + "\n"))
		}
		document[familyName] = bucket
	}

	return map[string]any{
		"digest":         hex.EncodeToString(hasher.Sum(nil)),
		"families":       document,
		"generator":      "go",
		"probe_count":    probeCount,
		"record_count":   recordCount,
		"schema_version": "freight-selftest/2",
		"series_count":   seriesCount,
	}
}
